File: app/src/main/java/com/example/getfood/ui/orderdetail/OrderDetailDisplayAdapter.java
package com.example.getfood.ui.orderdetail;

import android.support.annotation.NonNull;
import android.support.v7.recyclerview.extensions.ListAdapter;
import android.support.v7.util.DiffUtil;
import android.support.v7.widget.RecyclerView;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RatingBar;
import android.widget.TextView;

import com.example.getfood.R;
import com.fazemeright.canteen_app_models.models.CartItem;

import java.util.List;

public class OrderDetailDisplayAdapter extends ListAdapter<CartItem, OrderDetailDisplayAdapter.ViewHolder> {

    private OnOrderItemClickListener onOrderItemClickListener;

    OrderDetailDisplayAdapter(OnOrderItemClickListener onOrderItemClickListener) {
        super(new OrderDetailDiffCallBack());
        this.onOrderItemClickListener = onOrderItemClickListener;
    }

    void swapData(List<CartItem> newList) {
        submitList(newList);
    }

    @NonNull
    @Override
    public OrderDetailDisplayAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.order_display_layout, parent, false);
        return new ViewHolder(view);

    }

    @Override
    public void onBindViewHolder(@NonNull final OrderDetailDisplayAdapter.ViewHolder holder, int position) {

        CartItem item = getItem(position);

        if (item != null) {
            holder.bind(item);
        }
    }

    interface OnOrderItemClickListener {
        void onRatingGiven(float rating, int position);
    }

    static class OrderDetailDiffCallBack extends DiffUtil.ItemCallback<CartItem> {

        @Override
        public boolean areItemsTheSame(@NonNull CartItem cartItem, @NonNull CartItem t1) {
            return cartItem.getItemName().equals(t1.getItemName());
        }

        @Override
        public boolean areContentsTheSame(@NonNull CartItem cartItem, @NonNull CartItem t1) {
            return cartItem.equals(t1);
        }
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        TextView orderItemQuantityTextView, orderItemNameTextView, orderItemStatusTextView;
        RatingBar ratingBar;

        ViewHolder(View itemView) {
            super(itemView);
            orderItemQuantityTextView = itemView.findViewById(R.id.orderItemQuantityTextView);
            orderItemNameTextView = itemView.findViewById(R.id.orderItemNameTextView);
            orderItemStatusTextView = itemView.findViewById(R.id.orderItemStatusTextView);
            ratingBar = itemView.findViewById(R.id.ratingBar);
        }

        void bind(CartItem item) {
            if (item.getItemName() != null) {
                this.orderItemNameTextView.setText(item.getItemName());
            }
            if (item.getItemQuantity() != null) {
                orderItemQuantityTextView.setText(String.valueOf(item.getItemQuantity()));
            }
            if (item.getItemStatus() != null) {
                orderItemStatusTextView.setText(item.getItemStatus());
            }

            if (item.getItemRating() != 0) {
                ratingBar.setRating((float) item.getItemRating());
                ratingBar.setEnabled(false);
            } else {
                ratingBar.setRating(0);
                ratingBar.setEnabled(true);
            }

            ratingBar.setOnRatingBarChangeListener(new RatingBar.OnRatingBarChangeListener() {
                @Override
                public void onRatingChanged(RatingBar ratingBar, float rating, boolean fromUser) {
                    if (fromUser) {
                        Log.d("##DebugData", "Rating is " + ratingBar.getRating());
                        ratingBar.setEnabled(false);
                        onOrderItemClickListener.onRatingGiven(ratingBar.getRating(), getAdapterPosition());
                    }
                }
            });
        }
    }
}

File: app/src/main/java/com/example/getfood/ui/orderdetail/OrderDetailDisplayAdapter.kt
package com.example.getfood.ui.orderdetail

import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.RatingBar
import android.widget.RatingBar.OnRatingBarChangeListener
import android.widget.TextView
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.ListAdapter
import androidx.recyclerview.widget.RecyclerView
import com.example.canteen_app_models.models.CartItem
import com.example.getfood.R

class OrderDetailDisplayAdapter internal constructor(private val onOrderItemClickListener: OnOrderItemClickListener) : ListAdapter<CartItem, OrderDetailDisplayAdapter.ViewHolder>(OrderDetailDiffCallBack()) {
    fun swapData(newList: List<CartItem?>?) {
        submitList(newList)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
                .inflate(R.layout.order_display_layout, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val item = getItem(position)
        if (item != null) {
            holder.bind(item)
        }
    }

    internal interface OnOrderItemClickListener {
        fun onRatingGiven(rating: Float, position: Int)
    }

    internal class OrderDetailDiffCallBack : DiffUtil.ItemCallback<CartItem>() {
        override fun areItemsTheSame(cartItem: CartItem, t1: CartItem): Boolean {
            return cartItem.itemName == t1.itemName
        }

        override fun areContentsTheSame(cartItem: CartItem, t1: CartItem): Boolean {
            return cartItem == t1
        }
    }

    inner class ViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView) {
        private var orderItemQuantityTextView: TextView = itemView.findViewById(R.id.orderItemQuantityTextView)
        private var orderItemNameTextView: TextView = itemView.findViewById(R.id.orderItemNameTextView)
        private var orderItemStatusTextView: TextView = itemView.findViewById(R.id.orderItemStatusTextView)
        private var ratingBar: RatingBar = itemView.findViewById(R.id.ratingBar)
        fun bind(item: CartItem) {
            if (item.itemName != null) {
                orderItemNameTextView.text = item.itemName
            }
            if (item.itemQuantity != null) {
                orderItemQuantityTextView.text = item.itemQuantity.toString()
            }
            if (item.itemStatus != null) {
                orderItemStatusTextView.text = item.itemStatus
            }
            if (item.itemRating != 0L) {
                ratingBar.rating = item.itemRating.toFloat()
                ratingBar.isEnabled = false
            } else {
                ratingBar.rating = 0f
                ratingBar.isEnabled = true
            }
            ratingBar.onRatingBarChangeListener = OnRatingBarChangeListener { ratingBar, _, fromUser ->
                if (fromUser) {
                    Log.d("##DebugData", "Rating is " + ratingBar.rating)
                    ratingBar.isEnabled = false
                    onOrderItemClickListener.onRatingGiven(ratingBar.rating, adapterPosition)
                }
            }
        }

    }
}